packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,8 +11,806 @@ import 'picker.dart';
+      ),
+    );
+  }
+}
+
+class _CupertinoDatePickerDateState extends State<CupertinoDatePicker> {
+  int textDirectionFactor;
+  CupertinoLocalizations localizations;
+
+  // Alignment based on text direction. The variable name is self descriptive,
+  // however, when text direction is rtl, alignment is reversed.
+  Alignment alignCenterLeft;
+  Alignment alignCenterRight;
+
+  // The currently selected values of the picker.
+  int selectedDay;
+  int selectedMonth;
+  int selectedYear;
+
+  // The controller of the day picker. There are cases where the selected value
+  // of the picker is invalid (e.g. February 30th 2018), and this dayController
+  // is responsible for jumping to a valid value.
+  FixedExtentScrollController dayController;
+
+  // Estimated width of columns.
+  List<double> estimatedColumnWidths = new List<double>(_kColumnTypes);
+
+  @override
+  void initState() {
+    super.initState();
+    selectedDay = widget.initialDateTime.day;
+    selectedMonth = widget.initialDateTime.month;
+    selectedYear = widget.initialDateTime.year;
+
+    dayController = new FixedExtentScrollController(initialItem: selectedDay - 1);
+  }
+
+  // Estimate the minimum width that each column needs to layout its content.
+  double _getColumnWidth(_pickerColumnType columnType) {
+    String longestText = '';
+
+    if (columnType == _pickerColumnType.dayOfMonth) {
+      for (int i = 1 ; i <=31; i++) {
+        final String dayOfMonth = localizations.datePickerDayOfMonth(i);
+        if (longestText.length < dayOfMonth.length)
+          longestText = dayOfMonth;
+      }
+    } else if (columnType == _pickerColumnType.month) {
+      for (int i = 1 ; i <=12; i++) {
+        final String month = localizations.datePickerMonth(i);
+        if (longestText.length < month.length)
+          longestText = month;
+      }
+    } else if(columnType == _pickerColumnType.year)
+      longestText = localizations.datePickerYear(2018);
+    else
+      assert(false, 'column type is not appropriate');
+
+    final TextPainter painter = new TextPainter(
+      text: TextSpan(
+        style: DefaultTextStyle.of(context).style,
+        text: longestText,
+      ),
+      textDirection: Directionality.of(context),
+    );
+    painter.layout();
+
+    return painter.maxIntrinsicWidth;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    textDirectionFactor = Directionality.of(context) == TextDirection.ltr ? 1 : -1;
+    localizations = CupertinoLocalizations.of(context) ?? const DefaultCupertinoLocalizations();
+
+    alignCenterLeft = textDirectionFactor == 1 ? Alignment.centerLeft : Alignment.centerRight;
+    alignCenterRight = textDirectionFactor == 1 ? Alignment.centerRight : Alignment.centerLeft;
+
+    estimatedColumnWidths[_pickerColumnType.dayOfMonth.index] = _getColumnWidth(_pickerColumnType.dayOfMonth);
+    estimatedColumnWidths[_pickerColumnType.month.index] = _getColumnWidth(_pickerColumnType.month);
```

Comment:
Consider using a map instead of a list, since this semantically maps from the _pickerColumnType enum to doubles.

This would eliminate the risk of, eg index out of bounds errors.